File: HermiteInterpolation/SplineKnots/DeBoorKnotsGenerator.cs
using System;
using HermiteInterpolation.MathFunctions;
using HermiteInterpolation.Numerics;

using HermiteInterpolation.Shapes.SplineInterpolation;
using HermiteInterpolation.Utils;

namespace HermiteInterpolation.SplineKnots
{
    public class DeBoorKnotsGenerator : KnotsGenerator//,IDeBoor
    {
        protected virtual double[] MainDiagonal(int unknownsCount)
        {
            return MyArrays.InitalizedArray<double>(unknownsCount, 4);
        }

        protected virtual double[] UpperDiagonal(int unknownsCount)
        {
            return MyArrays.InitalizedArray<double>(unknownsCount, 1);
        }

        protected virtual double[] LowerDiagonal(int unknownsCount)
        {
            return MyArrays.InitalizedArray<double>(unknownsCount, 1);
        }

        protected virtual double[] RightSide(Func<int, double> rightSide, double h, double dfirst, double dlast,
            int unknownsCount)
        {
            //var unknownsCount = unknownsCount - 2;
            var rs = new double[unknownsCount];
            //var equation
            h = 3/h;
            rs[0] = h*(rightSide(2) - rightSide(0)) - dfirst;
            rs[unknownsCount - 1] = h*(rightSide(unknownsCount + 1) - rightSide(unknownsCount - 1)) - dlast;
            for (var i = 1; i < unknownsCount-1; i++)
            {
                rs[i] = h*(rightSide(i + 2) - rightSide(i));
            }
            return rs;
        }

        public override KnotMatrix GenerateKnots(SurfaceDimension uDimension, SurfaceDimension vDimension)
        {
            if (uDimension.KnotCount < 4 || vDimension.KnotCount < 4)
            {
                return new DirectKnotsGenerator(Function).GenerateKnots(uDimension, vDimension);
            }

            var values =new KnotMatrix(uDimension.KnotCount, vDimension.KnotCount);
            InitializeKnots(uDimension, vDimension, values);

            FillXDerivations(values);
            FillXYDerivations(values);
            FillYDerivations(values);
            FillYXDerivations(values);

            return values;
        }

        protected virtual void InitializeKnots(SurfaceDimension uDimension, SurfaceDimension vDimension, KnotMatrix values)
        {
            var uSize = Math.Abs(uDimension.Max - uDimension.Min)/(uDimension.KnotCount - 1);
            var vSize = Math.Abs(vDimension.Max - vDimension.Min)/(vDimension.KnotCount - 1);
            var u = uDimension.Min;

            // Init Z
            for (var i = 0; i < uDimension.KnotCount; i++, u += uSize)
                //Parallel.For(0,uDimension.KnotCount,i=>)
            {
                var v = vDimension.Min;
                for (var j = 0; j < vDimension.KnotCount; j++, v += vSize)
                {
                    var z = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Z, u, v);
                        //Function.Z(u,v); //Z(u, v);

                    values[i, j] = new Knot(u, v, z);
                }
            }
            // Init Dx
            var uKnotCountMin1 = uDimension.KnotCount - 1;
            for (var j = 0; j < vDimension.KnotCount; j++)
            {
                values[0, j].Dx = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dx, values[0, j].X,
                    values[0, j].Y); //Function.Dx(values[0,j].X, values[0,j].Y);
                values[uKnotCountMin1, j].Dx = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dx,
                    values[uKnotCountMin1, j].X,
                    values[uKnotCountMin1, j].Y);
            }
            // Init Dy
            var vKnotCountMin1 = vDimension.KnotCount - 1;
            for (var i = 0; i < uDimension.KnotCount; i++)
            {
                values[i, 0].Dy = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dy, values[i, 0].X,
                    values[i, 0].Y);
                values[i, vKnotCountMin1].Dy = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dy,
                    values[i, vKnotCountMin1].X,
                    values[i, vKnotCountMin1].Y);
            }
            // Init Dxy
            values[0, 0].Dxy = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dxy, values[0, 0].X,
                values[0, 0].Y);
            values[uKnotCountMin1, 0].Dxy = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dxy,
                values[uKnotCountMin1, 0].X,
                values[uKnotCountMin1, 0].Y);
            values[0, vKnotCountMin1].Dxy = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dxy,
                values[0, vKnotCountMin1].X,
                values[0, vKnotCountMin1].Y);
            values[uKnotCountMin1, vKnotCountMin1].Dxy = Numerics.MathFunctions.MathFunctionExtensions.SafeCall(Function.Dxy,
                values[uKnotCountMin1, vKnotCountMin1].X, values[uKnotCountMin1, vKnotCountMin1].Y);
        }

        protected virtual void FillXDerivations(KnotMatrix values)
        {
            for (var j = 0; j < values.Columns; j++)
            {
                FillXDerivations(j, values);
            }
        }

        protected virtual void FillXYDerivations(KnotMatrix values)
        {
            FillXYDerivations(0, values);
            FillXYDerivations(values.Columns - 1, values);
        }

        protected virtual void FillYDerivations(KnotMatrix values)
        {
            for (var i = 0; i < values.Rows; i++)
            {
                FillYDerivations(i, values);
            }
        }

        protected virtual void FillYXDerivations(KnotMatrix values)
        {
            for (var i = 0; i < values.Rows; i++)
            {
                FillYXDerivations(i, values);
            }
        }

        protected virtual void FillXDerivations(int columnIndex, KnotMatrix values)
        {
            var unknownsCount = values.Rows - 2;
            if (unknownsCount == 0) return;
            Action<int, double> dset = (idx, value) => values[idx,columnIndex].Dx = value;
            Func<int, double> rget = idx => values[idx,columnIndex].Z;
            var h = values[1,0].X - values[0,0].X;
            var dlast = values[values.Rows - 1,columnIndex].Dx;
            var dfirst = values[0,columnIndex].Dx;

            SolveTridiagonal(rget, h, dfirst, dlast, unknownsCount, dset);
        }



        protected virtual void FillYDerivations(int rowIndex, KnotMatrix values)
        {
            var unknownsCount = values.Columns - 2;
            if (unknownsCount == 2) return;
            Action<int, double> dset = (idx, value) => values[rowIndex,idx].Dy = value;
            Func<int, double> rget = idx => values[rowIndex,idx].Z;
            var h = values[0,1].Y - values[0,0].Y;
            var dfirst = values[rowIndex,0].Dy;
            var dlast = values[rowIndex,values.Columns - 1].Dy;

            SolveTridiagonal(rget, h, dfirst, dlast, unknownsCount, dset);
        }

        protected virtual void FillXYDerivations(int columnIndex, KnotMatrix values)
        {
            var unknownsCount = values.Rows - 2;
            if (unknownsCount == 2) return;
            Action<int, double> dset = (idx, value) => values[idx,columnIndex].Dxy = value;
            Func<int, double> rget = idx => values[idx,columnIndex].Dy;
            var h = values[1,0].X - values[0,0].X;
            var dlast = values[values.Rows - 1,columnIndex].Dxy;
            var dfirst = values[0,columnIndex].Dxy;

            SolveTridiagonal(rget, h, dfirst, dlast, unknownsCount, dset);
        }

        protected virtual void FillYXDerivations(int rowIndex, KnotMatrix values)
        {
            var unknownsCount = values.Columns - 2;
            if (unknownsCount == 2) return;
            Action<int, double> dset = (idx, value) => values[rowIndex,idx].Dxy = value;
            Func<int, double> rget = idx => values[rowIndex,idx].Dx;
            var h = values[0,1].Y - values[0,0].Y;
            var dfirst = values[rowIndex,0].Dxy;
            var dlast = values[rowIndex,values.Columns - 1].Dxy;

            SolveTridiagonal(rget, h, dfirst, dlast, unknownsCount, dset);
        }

        protected virtual void SolveTridiagonal(Func<int, double> rightSideValuesToGet, double h, double dfirst, double dlast,
            int unknownsCount, Action<int, double> unknownsToSet)
        {
            var result = RightSide(rightSideValuesToGet, h, dfirst, dlast, unknownsCount);
            LinearSystems.SolveTridiagonalSystem(UpperDiagonal(unknownsCount), MainDiagonal(unknownsCount),
                LowerDiagonal(unknownsCount), result);

            for (var i = 0; i < result.Length; i++)
            {
                unknownsToSet(i + 1, result[i]);
            }
        }

        public DeBoorKnotsGenerator(InterpolativeMathFunction function) : base(function)
        {
        }

        public DeBoorKnotsGenerator(MathExpression expression) : base(expression)
        {
        }
    }
}

File: SplineKnotsBenchmark/DeBoorKnotsGenerator.cpp
#include "stdafx.h"
#include "DeBoorKnotsGenerator.h"

#include "Tridiagonal.h"
#include <omp.h>
#include "SplineKnots.h"

#include <algorithm>

namespace splineknots
{
	Tridiagonals& DeBoorKnotsGenerator::Tridagonals()
	{
		return tridagonals_;
	}

	DeBoorKnotsGenerator::Precalculated::Precalculated(const double h)
		: h(h), three_div_h(3 / h)
	{
	}

	DeBoorKnotsGenerator::DeBoorKnotsGenerator(MathFunction math_function)
		: KnotsGenerator(math_function),
		tridagonals_(), is_parallel_(false)
	{
		tridagonals_.push_back(std::make_unique<splineknots::Tridiagonal>(1, 4, 1));
	}

	DeBoorKnotsGenerator::DeBoorKnotsGenerator(InterpolativeMathFunction math_function)
		: KnotsGenerator(math_function),
		tridagonals_(), is_parallel_(false)
	{
		tridagonals_.push_back(std::make_unique<splineknots::Tridiagonal>(1, 4, 1));
	}


	DeBoorKnotsGenerator::DeBoorKnotsGenerator(MathFunction math_function, std::unique_ptr<splineknots::Tridiagonal> tridiagonal)
		: KnotsGenerator(math_function),
		tridagonals_(), is_parallel_(false)
	{
		tridagonals_.push_back(std::move(tridiagonal));
	}

	DeBoorKnotsGenerator::DeBoorKnotsGenerator(InterpolativeMathFunction math_function, std::unique_ptr<splineknots::Tridiagonal> tridiagonal)
		: KnotsGenerator(math_function),
		tridagonals_(), is_parallel_(false)
	{
		tridagonals_.push_back(std::move(tridiagonal));
	}

	DeBoorKnotsGenerator::~DeBoorKnotsGenerator()
	{
	}

	DeBoorKnotsGenerator::DeBoorKnotsGenerator(const DeBoorKnotsGenerator& other) :
		KnotsGenerator(other),
		tridagonals_(other.tridagonals_.size()),
		is_parallel_(other.is_parallel_)
	{
		for (size_t i = 0; i < other.tridagonals_.size(); i++)
		{
			auto trid = std::unique_ptr<splineknots::Tridiagonal>(other.tridagonals_[i]->Clone());
			tridagonals_.push_back(std::move(trid));
		}
	}

	DeBoorKnotsGenerator::DeBoorKnotsGenerator(DeBoorKnotsGenerator&& other) :
		KnotsGenerator(std::move(other)),
		tridagonals_(std::move(other.tridagonals_)),
		is_parallel_(other.is_parallel_)
	{
	}

	DeBoorKnotsGenerator& DeBoorKnotsGenerator::operator=(const DeBoorKnotsGenerator& other)
	{
		if (this == &other)
			return *this;
		KnotsGenerator::operator =(other);
		tridagonals_.clear();
		for (size_t i = 0; i < other.tridagonals_.size(); i++)
		{
			auto trid = std::unique_ptr<splineknots::Tridiagonal>(other.tridagonals_[i]->Clone());
			tridagonals_.push_back(std::move(trid));
		}
		is_parallel_ = other.is_parallel_;
		return *this;
	}

	DeBoorKnotsGenerator& DeBoorKnotsGenerator::operator=(DeBoorKnotsGenerator&& other)
	{
		if (this == &other)
			return *this;
		KnotsGenerator::operator =(std::move(other));
		tridagonals_ = std::move(other.tridagonals_);
		is_parallel_ = other.is_parallel_;
		return *this;
	}

	KnotMatrix DeBoorKnotsGenerator::GenerateKnots(const SurfaceDimension& udimension, const SurfaceDimension& vdimension)
	{
		if (udimension.knot_count < 4 || vdimension.knot_count < 4)
		{
			return KnotMatrix::NullMatrix();
		}
		KnotMatrix values(udimension.knot_count, vdimension.knot_count);
		InitializeBuffers(udimension.knot_count, vdimension.knot_count);
		InitializeKnots(udimension, vdimension, values);
		FillXDerivations(values);
		FillXYDerivations(values);
		FillYDerivations(values);
		FillYXDerivations(values);
		return values;
	}

	void DeBoorKnotsGenerator::RightSide(const RightSideSelector& right_side_variables, const Precalculated& precalculated, const double dfirst, const double dlast,
		const int unknowns_count, double* rightside_buffer)
	{
		auto h3 = precalculated.three_div_h;
		rightside_buffer[0] = h3 * (right_side_variables(2) - right_side_variables(0)) - dfirst;
		rightside_buffer[unknowns_count - 1] = h3 * (right_side_variables(unknowns_count + 1) - right_side_variables(unknowns_count - 1)) - dlast;
		for (auto i = 1; i < unknowns_count - 1; i++)
		{
			rightside_buffer[i] = h3 * (right_side_variables(i + 2) - right_side_variables(i));
		}
	}

	void DeBoorKnotsGenerator::Precalculate(const SurfaceDimension& udimension, const SurfaceDimension& vdimension)
	{
		SetPrecalculatedHX(std::make_unique<Precalculated>(abs(udimension.max - udimension.min) / (udimension.knot_count - 1)));
		SetPrecalculatedHY(std::make_unique<Precalculated>(abs(vdimension.max - vdimension.min) / (vdimension.knot_count - 1)));
	}

	void DeBoorKnotsGenerator::InitializeKnots(const SurfaceDimension& udimension, const SurfaceDimension& vdimension, KnotMatrix& values)
	{
		Precalculate(udimension, vdimension);
		auto hx = precalculated_hx_->h;
		auto hy = precalculated_hy_->h;
		auto u = udimension.min;
		auto f = Function();
		// Init Z
		for (auto i = 0; i < udimension.knot_count; i++, u += hx)
		{
			auto v = vdimension.min;
			for (auto j = 0; j < vdimension.knot_count; j++, v += hy)
			{
				auto z = f.Z()(u, v);
				//Function.Z(u,v); //Z(u, v);
				values(i, j) = Knot(u, v, z);
			}
		}
		// Init Dx
		auto uKnotCountMin1 = udimension.knot_count - 1;
		for (auto j = 0; j < vdimension.knot_count; j++)
		{
			auto dx = f.Dx()(values(0, j).X(), values(0, j).Y());
			values(0, j).SetDx(dx); //Function.Dx(values(0,j).X, values(0,j).Y);
			values(uKnotCountMin1, j).SetDx(f.Dx()(values(uKnotCountMin1, j).X(), values(uKnotCountMin1, j).Y()));
		}
		// Init Dy
		auto vKnotCountMin1 = vdimension.knot_count - 1;
		for (auto i = 0; i < udimension.knot_count; i++)
		{
			values(i, 0).SetDy(f.Dy()(values(i, 0).X(), values(i, 0).Y()));
			values(i, vKnotCountMin1).SetDy(
				f.Dy()(values(i, vKnotCountMin1).X(), values(i, vKnotCountMin1).Y())
				);
		}
		// Init Dxy
		values(0, 0).SetDxy(f.Dxy()(values(0, 0).X(), values(0, 0).Y()));
		values(uKnotCountMin1, 0).SetDxy(f.Dxy()(values(uKnotCountMin1, 0).X(), values(uKnotCountMin1, 0).Y()));
		values(0, vKnotCountMin1).SetDxy(f.Dxy()(values(0, vKnotCountMin1).X(), values(0, vKnotCountMin1).Y()));
		values(uKnotCountMin1, vKnotCountMin1).SetDxy(f.Dxy()(values(uKnotCountMin1, vKnotCountMin1).X(), values(uKnotCountMin1, vKnotCountMin1).Y()));
	}

	void DeBoorKnotsGenerator::FillXDerivations(KnotMatrix& values)
	{
		utils::For(0, static_cast<int>(values.ColumnsCount()),
			[&](int j)
		{
			FillXDerivations(j, values);
		},
			1, is_parallel_);
	}

	void DeBoorKnotsGenerator::FillXYDerivations(KnotMatrix& values)
	{
		FillXYDerivations(0, values);
		FillXYDerivations(values.ColumnsCount() - 1, values);
	}

	void DeBoorKnotsGenerator::FillYDerivations(KnotMatrix& values)
	{
		utils::For(0, static_cast<int>(values.RowsCount()),
			[&](int i)
		{
			FillYDerivations(i, values);
		},
			1, is_parallel_);
	}

	void DeBoorKnotsGenerator::FillYXDerivations(KnotMatrix& values)
	{
		utils::For(0, static_cast<int>(values.RowsCount()),
			[&](int i)
		{
			FillYXDerivations(i, values);
		},
			1, is_parallel_);
	}

	void DeBoorKnotsGenerator::FillXDerivations(const int column_index, KnotMatrix& values)
	{
		auto unknowns_count = values.RowsCount() - 2;
		if (unknowns_count == 0) return;

		UnknownsSetter dset = [&](int index, double value)
		{
			values[index][column_index].SetDx(value);
		};
		RightSideSelector rget = [&](int index)
		{
			return values[index][column_index].Z();
		};

		auto dlast = values(values.RowsCount() - 1, column_index).Dx();
		auto dfirst = values(0, column_index).Dx();

		SolveTridiagonal(rget, PrecalculatedHX(), dfirst, dlast, unknowns_count, dset);
	}

	void DeBoorKnotsGenerator::FillXYDerivations(const int column_index, KnotMatrix& values)
	{
		auto unknowns_count = values.RowsCount() - 2;
		if (unknowns_count == 0) return;

		UnknownsSetter dset = [&](int index, double value)
		{
			values[index][column_index].SetDxy(value);
		};
		RightSideSelector rget = [&](int index)
		{
			return values(index, column_index).Dy();
		};

		auto h = precalculated_hx_->h;
		auto dlast = values(values.RowsCount() - 1, column_index).Dxy();
		auto dfirst = values(0, column_index).Dxy();

		SolveTridiagonal(rget, PrecalculatedHX(), dfirst, dlast, unknowns_count, dset);
	}

	void DeBoorKnotsGenerator::FillYDerivations(const int row_index, KnotMatrix& values)
	{
		auto unknowns_count = values.ColumnsCount() - 2;
		if (unknowns_count == 0) return;

		UnknownsSetter dset = [&](int index, double value)
		{
			values[row_index][index].SetDy(value);
		};
		RightSideSelector rget = [&](int index)
		{
			return values(row_index, index).Z();
		};

		auto dlast = values(row_index, values.ColumnsCount() - 1).Dy();
		auto dfirst = values(row_index, 0).Dy();

		SolveTridiagonal(rget, PrecalculatedHY(), dfirst, dlast, unknowns_count, dset);
	}

	void DeBoorKnotsGenerator::FillYXDerivations(const int row_index, KnotMatrix& values)
	{
		auto unknowns_count = values.ColumnsCount() - 2;
		if (unknowns_count == 0) return;

		UnknownsSetter dset = [&](int index, double value)
		{
			values[row_index][index].SetDxy(value);
		};
		RightSideSelector rget = [&](int index)
		{
			return values(row_index, index).Dx();
		};

		auto dlast = values(row_index, values.ColumnsCount() - 1).Dxy();
		auto dfirst = values(row_index, 0).Dxy();

		SolveTridiagonal(rget, PrecalculatedHY(), dfirst, dlast, unknowns_count, dset);
	}

	bool DeBoorKnotsGenerator::IsParallel()
	{
		return is_parallel_;
	}

	const DeBoorKnotsGenerator::Precalculated& DeBoorKnotsGenerator::PrecalculatedHX() const
	{
		return *precalculated_hx_;
	}

	const DeBoorKnotsGenerator::Precalculated& DeBoorKnotsGenerator::PrecalculatedHY() const
	{
		return *precalculated_hy_;
	}

	void DeBoorKnotsGenerator::SetPrecalculatedHX(std::unique_ptr<Precalculated> precalculated)
	{
		precalculated_hx_ = std::move(precalculated);
	}

	void DeBoorKnotsGenerator::SetPrecalculatedHY(std::unique_ptr<Precalculated> precalculated)
	{
		precalculated_hy_ = std::move(precalculated);
	}

	void DeBoorKnotsGenerator::SolveTridiagonal(const RightSideSelector& selector, const Precalculated& precalculated, const double dfirst, const double dlast, const int unknowns_count, UnknownsSetter& unknowns_setter)
	{
		auto& tridiagonal = Tridiagonal(omp_get_thread_num());
		auto rightside = tridiagonal.RightSideBuffer();
		RightSide(selector, precalculated, dfirst, dlast, unknowns_count, rightside);

		tridiagonal.Solve(unknowns_count);
		for (int k = 0; k < unknowns_count; k++)
		{
			unknowns_setter(k + 1, rightside[k]);
		}
	}

	void DeBoorKnotsGenerator::InParallel(bool value)
	{
		is_parallel_ = value;
		auto threads = utils::num_threads;
		if (value)
		{
			tridagonals_.reserve(threads);
			for (auto i = tridagonals_.size(); i < threads; i++)
			{
				// create copy of tridiagonal solver
				std::unique_ptr<splineknots::Tridiagonal> copy_of_first(tridagonals_[0]->Clone());
				tridagonals_.push_back(std::move(copy_of_first));
			}
		}
		else
		{
			tridagonals_._Pop_back_n(tridagonals_.size() - 1);
		}
	}

	void DeBoorKnotsGenerator::InitializeBuffers(const size_t u_count, const size_t v_count)
	{
		auto size = std::max(u_count - 2, v_count - 2);
		for (size_t i = 0; i < tridagonals_.size(); i++)
		{
			tridagonals_[i]->ResizeBuffers(size);
		}
	}

	splineknots::Tridiagonal& DeBoorKnotsGenerator::Tridiagonal(int index)
	{
		return *tridagonals_[index];
	}
}
